### ISSUE TITLE
abcd letters as separate keys for mapping house numbers

### ISSUE THREAD

**Issue by canfe** [streetcomplete/StreetComplete#5479]
**Use case**
When mapping house numbers that have an "alphabetic extension" such as: 45b, it is not convenient to use the full keyboard to look up the latin letter you need.
**Proposed Solution**
It would be helpful to have the first 3-4 letters of the alphabet highlighted as separate keys. This would solve the great majority of cases.

**Comment by westnordost** [comment]
I strive to have the UI as clean as possible, i.e. be very cautious about adding any buttons that are not necessary. I see several reasons for not doing this:

1. Adding 3-4 buttons to make it marginally faster to answer the housenumber question for an estimated 5% of housenumbers doesn't seem proportionate to me. An other case where such quick select buttons exist is the building levels quest, but there, one uses these button in the vast majority of cases, I'd guess 90% of the time.

2. These [a] [b] [c] [d] buttons would roughly have the same size of the same buttons on the keyboard, which is already displayed. So, it really just saves one tap on the [ABC] button and if the user already tapped on that button, it is a duplication

3. At least for certain countries, the form is actually quite stuffed already, e.g. in Japan where addresses also have an associated block number or Slovakia where addresses may also have a conscription number.

**Comment by canfe** [comment]
> * What would work ideally for me personally, is to have `+` / `-` buttons which work as they do now (increase/decrease the number), but also have `a+` and `a-` buttons which would increase/decrease letters (adding or removing them completely as needed). 


**great idea!**

**Comment by thiemowmde** [comment]
I'm currently mapping a German city with a **lot** of house numbers like 45b. A small improvement that would save one or two extra button presses each time is to automatically open the full keyboard instead of the numeric one when the previous house number ends with a letter.

**Comment by westnordost** [comment]
In other words, make the ABC-button appear and work like a [toggle button](https://m2.material.io/components/buttons#toggle-button), memorize and re-apply the previous state of it when next opening the form.
This sounds good, a PR for that would be accepted.

@mnalis suggestion sounds fine, but I am not sure if it fits and/or if it wouldn't confuse users. The buttons should probably not be titled a- / a+ though, but bear the actual letter it would apply to the housenumber. (E.g. for housenumber 43c, the buttons would show [b] and [d])

**Comment by westnordost** [comment]
In other words, make the ABC-button appear and work like a [toggle button](https://m2.material.io/components/buttons#toggle-button), memorize and re-apply the previous state of it when next opening the form.
This sounds good, however, this means it is necessary to switch back when the ABC-numbering changes to the next housenumber. Do we have some numbers? How likely is it that if a housenumber with an "a" is encountered, there will also be a "b"?
If that probability is less than 50% or around 50%, this suggestion would actually make the UI worse.

@mnalis suggestion sounds fine, but I am not sure if it fits and/or if it wouldn't confuse users. The buttons should probably not be titled a- / a+ though, but bear the actual letter it would apply to the housenumber. (E.g. for housenumber 43c, the buttons would show [b] and [d]).

**Comment by canfe** [comment]
@westnordost If housenumber 43c, the buttons would show [b] and [d], **but if the next number is pure 45**, it would be functional to click on the previous displayed number, i.e., 43c and then on the + button to increase only the digits without the Latin characters.
Basically the layout would not be changed, only the functionality.

**Comment by mnalis** [comment]
> Do we have some numbers? How likely is it that if a housenumber with an "a" is encountered, there will also be a "b"?

Quick [calculation](https://gist.github.com/mnalis/0e808d847ccad8597a010e127bd5b889) shows the global distribution of `b` housenumbers compared to `a` housenumbers is about `41%`.  
(I'ts not really precise, because it doesn't really look at closeby houses but just at numbers followed by letters. And it probably significantly differs by countries, so people can run the script on their own country to get the idea)

> @mnalis suggestion sounds fine, but I am not sure if it fits and/or if it wouldn't confuse users. The buttons should probably not be titled a- / a+ though, but bear the actual letter it would apply to the housenumber. (E.g. for housenumber 43c, the buttons would show [b] and [d]).

Yes, that sounds like better idea to me:
- when housenumber is  e.g. `43` it would show only next-button `a` (to advance to `43a`)
- when housenumber is  e.g. `43z` it would show only prev-button `y` (to decrease to `43y`)
- when housenumber is ending in any other letter e.g. `43d`, it would show both prev-button `43c` and next-button `43e`.

And it would only cost space for two buttons while covering majority of usages. Something like this:
![housenumber_letters_s](https://github.com/streetcomplete/StreetComplete/assets/156656/676cb49b-b233-4ad3-b55f-000b528d3ef3)

Or maybe even move `+` / `-` buttons on the left side of housenumber to reduce misclicks?

**Comment by westnordost** [comment]
Thank you for the numbers! By the way, I think this is potentially a use case in which sophox or qlever could provide an answer within (milli)seconds.

That's quite a lot of buttons, though. Hmm. How would that look like in Slowakia and other regions where the housenumber layout takes up more space? Anyway, to put the +/- on the left side in case this is implemented would make sense. The ABC-button should probably also be replaced with a dedicated icon showing a keyboard (+ABC in small) to reduce confusion.

**Comment by mnalis** [comment]
> Thank you for the numbers! By the way, I think this is potentially a use case in which sophox or qlever could provide an answer within (milli)seconds.

Perhaps, but it would take me significantly more time to learn SPARQL then milliseconds :smiley: 

> How would that look like in Slowakia and other regions where the housenumber layout takes up more space?

https://github.com/streetcomplete/StreetComplete/blob/e1d19abdd84c77155894b4fe8526cc71e0e42396/app/src/main/java/de/westnordost/streetcomplete/quests/address/AddHousenumberForm.kt#L186-L190 mentions 3 countries with separate form:


- In **Slovakia** in particular, the current UI in SC does not seem to implement even current `+` / `-` to increment/decrement housenumbers (nor does it remember previous value, instead opening with blank form on every new house), so I guess they would not get buttons for house-letters either. 

- **Czechia** has HousenumberQuest [disabled](https://github.com/streetcomplete/StreetComplete/blob/e1d19abdd84c77155894b4fe8526cc71e0e42396/app/src/main/java/de/westnordost/streetcomplete/quests/address/AddHousenumber.kt#L39), but Address Overlay is similar as in Slovakia, without `+`/`-` buttons.
- **Japan** looks mostly similar to general case, but has an additional block number, so it should be easy to add too. However, I do [not see housenumbers ending with alphabetic letters there](https://overpass-turbo.eu/s/1HBR), so I'm not sure if would be worthwhile implementing it in that form too (perhaps someone with more experience with Japan will chip in).

The Slovakia & Japan forms currently look like this:
  ![small_Screenshot_20240220_183053_StreetComplete](https://github.com/streetcomplete/StreetComplete/assets/156656/59b1af9d-9c1d-4831-b26b-d4d9f99fc28a)  ![small_Screenshot_20240220_182204_StreetComplete](https://github.com/streetcomplete/StreetComplete/assets/156656/c465db7e-735e-4d02-9cf0-39b1b4039449)

---

**TL;DR:** I think that just the general form would need improving (at least for now). Mockup could look something like this (keyboard icon taken from opening hours quest):

![housenumber_letters_s](https://github.com/streetcomplete/StreetComplete/assets/156656/4fbf7f23-c250-4fd5-9a88-bffc2b344f2f)

**Comment by rhhsm** [comment]
> Japan looks mostly similar to general case, but has an additional block number, so it should be easy to add too. However, I do [not see housenumbers ending with alphabetic letters there](https://overpass-turbo.eu/s/1HBR), so I'm not sure if would be worthwhile implementing it in that form too (perhaps someone with more experience with Japan will chip in).

If they would distinguish households sharing the same house&block number, they would be unlikely to use Latin script for it :) But they don't: households sharing the same address are distinguished by the family name on the letterbox. So the UI for Japan can be kept as it is.

**Comment by rhhsm** [comment]
Looking forward to try if this will also work for Cyrillic script (А Б В Г...)

**Comment by westnordost** [comment]
It should, actually, but the implementation is rather naive: It just selects the next code point. This presupposes that the order of all the letters in unicode are actually ordered in alphabetical order, without any extra-letters in-between (such as umlauts öäü, ß, diacritics like `ñ, ´, ...) and terminating in a non-letter.

**Comment by rhhsm** [comment]
It should work fine for Bulgarian up to the 31st letter, but other Cyrillic alphabets have a special letter inserted at the 5th (Ukrainian), 6th (Serbian and Macedonian) or 7th (Russian, Belorussian) position that only comes much later in the unicode table. They will need to switch to keyboard to enter it.
